controllers/deleteProductControllerIntegration.test.js

```diff
@@ -0,0 +1,93 @@
+import mongoose from "mongoose";
+import { MongoMemoryServer } from "mongodb-memory-server";
+import productModel from "../models/productModel";
+import {deleteProductController} from "./productController";
+import fs from 'fs';
+
+let mongoServer;
+
+describe('integration of delete product controller', () => {
+    let req, res;
+
+    beforeAll(async () => {
+        mongoServer = await MongoMemoryServer.create();
+        const uri = mongoServer.getUri();
+        await mongoose.connect(uri);
+    });
+
+    beforeEach(() => {
+        req = {};
+        res = {
+            status: jest.fn().mockReturnThis(),
+            send: jest.fn(),
+        };
+    });
+
+    afterAll(async () => {
+        await mongoose.disconnect();
+        await mongoServer.stop();
+    });
+
+    test.failing('returns 200, success message and deletes product with valid input', async () => {
+        
+        //fails as original message is buggy (capitalisation)
+
+        //ARRANGE
+        const expectedBuffer = fs.readFileSync('client/public/images/test-pdt-img-1.jpg');
+        const expectedCategory = new mongoose.Types.ObjectId();
+
+        //products in the database
+        const pdt1 = await new productModel({name: "sneakers", slug: "sneakers", description: "adidas sneakers",price: 134, quantity: 10,
+            category: expectedCategory,  
+            photo: {
+              data: expectedBuffer,  
+              contentType: "image/png"
+            },
+            shipping: true, timestamp:Date.now()})
+        .save();
+        const pid1 = pdt1._id;
+
+        const pdt2 = await new productModel({name: "timeless", slug: "timeless", description: "book called timeless",price: 149, quantity: 71,
+            category: new mongoose.Types.ObjectId(),  
+                photo: {
+                    data: expectedBuffer,  
+                    contentType: "image/png"
+                },
+            shipping: true, timestamp:Date.now()})
+        .save();
+        const delPid = pdt2._id;
+
+        req.params = {
+            pid: delPid
+        };
+
+        //ACT
+        await deleteProductController(req, res);
+
+        //ASSERT
+        expect(res.status).toHaveBeenCalledWith(200);
+        expect(res.send).toHaveBeenCalledWith(expect.objectContaining({
+            success: true,
+            message: "Product Deleted Successfully",
+            })
+        );
+
+        //not deleted
+        const samplePdt1 = await productModel.findById(pid1);
```

Comment:
Is it necessary to check that product 1 is not deleted? Since this test is testing deletion of product 2, it seems not necessary to check whether product 1 is deleted or not...? This is because the side-effect of accidentally deleting another product via deleteProduct controller is not likely to occur?

Reply:
Thanks for the suggestion! What I am trying to assert here is that the delete product correctly deletes intended project _only_. I guess it depends on how we choose to scope the tests, but in this case, this assertion helps check the same. An example to think of would be an underlying implementation/issue with integration that deletes all products. In this case, such a bug would still pass assertion that intended product is deleted but would fail this other one.